rfcs/2020-04-06-1858-automatically-adjust-limits.md

```diff
@@ -0,0 +1,164 @@
+* [Additive Increase/Multiplicative Decrease](https://en.wikipedia.org/wiki/Additive_increase/multiplicative_decrease)
+* [Netflix Technology Blog: Performance Under Load](https://medium.com/@NetflixTechBlog/performance-under-load-3e6fa9a60581)
+* [JINSPIRED - Adaptive Safety Control (archive.org)](https://web.archive.org/web/20130105023839/http://www.jinspired.com/site/jxinsight-opencore-6-4-ea-11-released-adaptive-safety-control)
+
+## Sales Pitch
+
+This proposal:
+
+* provides a simple and understandable mechanism for varying resource
+  utilization of sinks;
+
+* adapts an existing design to avoid reinventing known good solutions;
+
+* is minimally invasive to the existing code base while applying to most
+  sinks;
+
+* minimizes the amount of configuration required to produce the ideal
+  (most efficient and performant) configuration; and
+
+* does not impose hard limits on flow rates while respecting configured
+  limits.
+
+## Drawbacks
+
+Since the underlying parameters that control when requests are throttled
+will be abstracted behind an additional layer, it will become harder to
+reason about the causes of bandwidth limits.
+
+## Rationale
+
+* As referenced earlier, the proposed mechanism borrows from _proven_
+  mechanisms designed to manage flow control under varying conditions,
+  making it a good choice for the first pass implementation.
+
+* A moving average is used to smooth out small variations in latency
+  without completely ignoring them.
+
+* EWMA is chosen as an averaging mechanism as it avoids having to
+  maintain memory of past observations beyond a single
+  value. Mathematically it is the simplest possible moving average.
+
+## Alternatives
+
+* Instead of managing the concurrency, we could alter the maximum
+  request rate or maximum bandwidth usage. This runs into the difficulty
+  of how to set the minimum bound both before any data has been
+  collected and after hard backoffs, while concurrency has a trivially
+  obvious minimum bound and is better able to flex with load.
+
+* Instead of comparing the RTT against a moving average, we could simply
+  use the previous observation (mathematically equivalent to a EWMA
+  weighting of α=1).
+
+## Outstanding Questions
+
+* The ideal value for the weighting α is unknown. Too large a value will
+  amplify the effect of short term changes to RTT. Too small a value may
+  delay responding to real changes excessively.
+
+* Some experimentation may be required to determine a small zone around
+  the average that is still considered "equal" to avoid excessive
+  flapping of the concurrency level without allowing the RTT to grow
+  unbounded and overload the sink.
+
+* Some level of (random) jitter may be needed to stagger the increases,
+  to avoid a large number of clients overwhelming a sink.
+
+## Plan Of Attack
+
+* [ ] Submit a PR with spike-level code _roughly_ demonstrating the change.
+* [ ] Benchmark the approach under various conditions to determine a good
```

Comment:
This, I think, needs much more detail. How exactly do you plan to test/benchmark this change to ensure it is:

1. Behaving as expected.
2. Does not regress over time.
3. Allows us to expand as we learn more.

For example. Starting simple, it would help to define a set of scenarios and their ideal outcome. We could then replay these scenarios to verify behavior as we hone in on the optimal algorithm. In other words, tests 😄 . But I think we should agree on those test criteria within this RFC. Ex:

1.  Service suddenly becomes unresponsive with sustained request time outs.
    * Vector should eventually reduce to an `in_flight_limit` of `1`.
2.  Service gradually increases in response time by 100ms each request.
    * Vector should eventually reduce to an `in_flight_limit` of `1`.
3.  Service implements a rate limit of 5 requests per second. If this limit is exceeded a `429` is returned:
    * Vector's `in_flight_limit` should hover around `5`.
4. etc..

It would be nice to have these static expectations to keep us honest. Beyond that, I think we would benefit from a randomized simulation of some sort. For example, we could define criteria that would cause a mock service to become unresponsive (shut down) and the test fails if the service reaches that state. The service would implement a rate-limit that randomly adjusts, and if the algorithm works, Vector should correctly hone in on that rate-limit as it adjusts. This should all be done within defined bounds that, if exceeded, would shut down the service.

Easier said than done! I'm happy to explain better. And I have to assume there is prior art for this kind of testing. If not, it would make for excellent blog post!

Reply:
I agree this needs a more concrete testing and benchmarking plan, and these are all good ideas. Does all of this detail really belong in this implementation plan, or am I free to spread it elsewhere and just reference it here?

Reply:
Not at all. Feel free to add a section or add it under the "Guide Level Proposal" section.

Reply:
And to clarify the "Plan of Attach", it should be progressive and outline an optimal plan. It does not mean we will do all of the items. For example, if a simulator is part of the plan, I would expect that to be at the end unless it's absolutely necessary. It's possible we won't do it depending on how everything comes together.

Reply:
BTW condition 3 above doesn't actually compute—the rate limit and concurrency are different units. A rate limit of 5 requests per second might be achieved with any number of different concurrency levels. It depends on what the RTT is between Vector and the service. More precisely, Vector's concurrency should approximate `rate_limit / RTT`.